File: src/poscar_2ctrls.cpp
#include <CLI/CLI.hpp>
#include <iostream>
#include <string>

#include "poscar_file.h"

int main(int argc, char* argv[]) {
    CLI::App app{"Transform POSCAR structure format to ctrls format used by ecalj/Qestaal"};

    std::string inputFile{"POSCAR"};
    std::string outputFile{"ctrls.system"};

    app.add_option("--input,-i", inputFile, "Input POSCAR file")->capture_default_str()->check(CLI::ExistingFile);
    app.add_option("--output,-o", outputFile, "Output ctrls file")->capture_default_str();

    CLI11_PARSE(app, argc, argv);

    POSCAR poscar;
    if (!poscar.readPOSCAR(inputFile)) {
        std::cerr << "Error: failed to parse input POSCAR " << inputFile << "\n";
        return 1;
    }

    if (poscar.is_direct) {
        poscar.toCartesian();
        std::cout << "Converted Direct -> Cartesian\n";
    }

    if (!poscar.writeCtrlsFile(outputFile)) {
        std::cerr << "Error: writing output ctrls file " << outputFile << "\n";
        return 1;
    }

    std::cout << "Output written to: " << outputFile << "\n";

    return 0;
}

File: rust/src/bin/poscar_2ctrls.rs
use clap::Parser;
use vasp_utils::poscar::Poscar;

#[derive(Parser)]
#[command(about = "Convert POSCAR to ecalj/Questaal ctrls format (Å → Bohr)")]
struct Cli {
    #[arg(short, long, default_value = "POSCAR")]
    input: String,
    #[arg(short, long, default_value = "ctrls.system")]
    output: String,
}

fn main() {
    let cli = Cli::parse();

    let mut poscar = match Poscar::read(&cli.input) {
        Ok(p) => p,
        Err(e) => {
            eprintln!("{e}");
            std::process::exit(1);
        }
    };

    if poscar.is_direct {
        poscar.to_cartesian();
        println!("Converted Direct -> Cartesian");
    }

    if let Err(e) = poscar.write_ctrls(&cli.output) {
        eprintln!("{e}");
        std::process::exit(1);
    }
    println!("Output written to: {}", cli.output);
}
